platform/security/realm/security-realm-userpass/src/main/java/org/codice/ddf/security/userpass/realm/UsernamePasswordRealm.java

```diff
@@ -0,0 +1,227 @@
+      throws AuthenticationException {
+    String credentials = (String) token.getCredentials();
+    String[] userpass = credentials.split(":");
+    if (userpass.length != 2) {
+      throw new AuthenticationException("Credentials were not in the correct format.");
+    }
+    String user = new String(Base64.getDecoder().decode(userpass[0]), StandardCharsets.UTF_8);
+    String pass = new String(Base64.getDecoder().decode(userpass[1]), StandardCharsets.UTF_8);
+    Subject subject = null;
+    for (JaasRealm jaasRealm : realmList) {
+      try {
+        subject = login(user, pass, jaasRealm.getName());
+        LOGGER.trace("Login succeeded for {} against realm {}", user, jaasRealm.getName());
+        break;
+      } catch (LoginException e) {
+        LOGGER.trace("Login failed for {} against realm {}", user, jaasRealm.getName());
+      }
+    }
+    if (subject != null) {
+      SimpleAuthenticationInfo simpleAuthenticationInfo = new SimpleAuthenticationInfo();
+      SimplePrincipalCollection principalCollection = createPrincipalCollectionFromSubject(subject);
+      simpleAuthenticationInfo.setPrincipals(principalCollection);
+      simpleAuthenticationInfo.setCredentials(credentials);
+      return simpleAuthenticationInfo;
+    }
+    throw new AuthenticationException("Login failed for user: " + user);
+  }
+
+  private SimplePrincipalCollection createPrincipalCollectionFromSubject(Subject subject) {
+    SimplePrincipalCollection principals = new SimplePrincipalCollection();
+    DefaultSecurityAssertionBuilder assertionBuilder = new DefaultSecurityAssertionBuilder();
+
+    AttributeStatement attributeStatement = new AttributeStatementDefault();
+    Principal userPrincipal =
+        subject
+            .getPrincipals()
+            .stream()
+            .filter(p -> p instanceof UserPrincipal)
+            .findFirst()
+            .orElseThrow(AuthenticationException::new);
+    Set<Principal> rolePrincipals =
+        subject
+            .getPrincipals()
+            .stream()
+            .filter(p -> p instanceof RolePrincipal)
+            .collect(Collectors.toSet());
+    for (ClaimsHandler claimsHandler : claimsHandlers) {
+      ClaimsCollection claims =
+          claimsHandler.retrieveClaims(
+              new ClaimsParametersImpl(userPrincipal, rolePrincipals, new HashMap<>()));
+      mergeClaimsToAttributes(attributeStatement, claims);
+    }
+    final Instant now = Instant.now();
+
+    assertionBuilder
+        .addAttributeStatement(attributeStatement)
+        .userPrincipal(userPrincipal)
+        .weight(SecurityAssertion.LOCAL_AUTH_WEIGHT)
+        .issuer("DDF")
+        .notBefore(Date.from(now))
+        .notOnOrAfter(Date.from(now.plus(fourHours)));
+
+    for (Principal principal : rolePrincipals) {
+      assertionBuilder.addPrincipal(principal);
+    }
+
+    SecurityAssertion assertion = assertionBuilder.build();
+
+    principals.add(assertion, "UP");
```

Comment:
❔ Is this worth defining as a constant?

Reply:
It isn't referenced in any way, so should be fine. It just needs to be non-null.